modal/_asgi.py

```diff
@@ -15,12 +15,62 @@
 FIRST_MESSAGE_TIMEOUT_SECONDS = 5.0
 
 
-def asgi_app_wrapper(asgi_app, function_io_manager) -> Callable[..., AsyncGenerator]:
+class LifespanManager:
+    def __init__(self, asgi_app, state):
+        self.asgi_app = asgi_app
+        self.queue: Optional[asyncio.Queue[Dict[str, Any]]] = None
+        self.startup_complete: Optional[asyncio.Future[None]] = None
+        self.state = state
+
+    async def background_task(self):
+        self.queue = asyncio.Queue()
+        self.startup: asyncio.Future[None] = asyncio.Future()
+        self.shutdown: asyncio.Future[None] = asyncio.Future()
+
+        async def receive():
+            if self.queue is None:
+                raise ValueError("queue is not initialized, call background_task first")
+            return await self.queue.get()
+
+        async def send(message):
+            if message["type"] == "lifespan.startup.complete":
+                self.startup.set_result(None)
+            elif message["type"] == "lifespan.startup.failed":
+                self.startup.set_exception(Exception("Startup failed"))
```

Comment:
Here we might want to be more explicit, e.g. "ASGI lifespan startup failed".

Also - do we have a test for this?

Reply:
Added tests for startup & shutdown failures